sympy/solvers/ode.py

```diff
@@ -346,6 +346,160 @@
     )
 
 
+class Solver:
+    def __init__(self, eq, func, hint="default", simplify=True, ics=None):
+        # Store the inputs passed to dsolve
+        self.eq, self.func = self.pre_process(eq, func)
+        self.x = self.func.args[0]
+        self.order = ode_order(eq, self.func.func(self.x))
+        self.ics = ics
+        self.simplify = simplify
+        self.hint = hint
+
+    def pre_process(self, eq, func):
+        # To convert equality to expression and find function f if not given by user.
+        # Same as that used by _desolve()
+        if isinstance(eq, Equality):
+            eq = eq.lhs - eq.rhs
+        if isinstance(eq, Pow):
+            if (eq.exp).is_positive:
+                eq = eq.base
+        derivs = eq.atoms(Derivative)
+        if not func:
+            funcs = set().union(*[d.atoms(AppliedUndef) for d in derivs])
+            if len(funcs) != 1:
+                raise ValueError('The function cannot be '
+                    'automatically detected for %s.' % eq)
+            func = funcs.pop()
+        return eq, func
+
+class FirstLinear(Solver):
+    def _get_coeffs_linear_ode(self):
+        order = self.order
```

Comment:
The idea of the `_get_coeffs_linear_ode` function is that it can be used by any of the solvers so it shouldn't be part of this class.